00_Base/src/util/command.executor.ts

```diff
@@ -360,4 +500,80 @@ export class CommandExecutor {
     );
     return setChargingProfileRequest;
   }
+
+  /**
+   * Create and store reservation related objects, including the core reservation, the ocpi reservation and the
+   * reserveNow request
+   *
+   * @param reserveNow - OCPI reserveNow
+   * @param evse - OCPI evse
+   * @param countryCode - MSP country code
+   * @param partyId - MSP party id
+   * @returns [reserveNowRequest, coreReservationDatabaseId]
+   */
+  private async createAndStoreReservations(
+    reserveNow: ReserveNow,
+    evse: OcpiEvse,
+    countryCode: string,
+    partyId: string,
+  ): Promise<[ReserveNowRequest, number]> {
+    const existingOcpiReservation =
+      await this.ocpiReservationRepo.readOnlyOneByQuery({
+        where: {
+          // unique constraint
+          [OcpiReservationProps.reservationId]: reserveNow.reservation_id,
+          [OcpiReservationProps.countryCode]: countryCode,
+          [OcpiReservationProps.partyId]: partyId,
+        },
+        include: [Reservation],
+      });
+
+    let coreReservationId;
+    if (!existingOcpiReservation) {
+      // Based on OCPI 2.1.1, The reservation_id sent by the Sender (eMSP) to the Receiver (CPO) SHALL NOT be sent
+      // directly to a Charge Point. The CPO SHALL make sure the Reservation ID sent to the Charge Point is unique and
+      // is not used by another Sender(eMSP).
+      coreReservationId = await this.reservationRepo.getNextReservationId(
+        evse.stationId,
+      );
+    } else {
+      coreReservationId = existingOcpiReservation.coreReservation.id;
+    }
+    if (!coreReservationId) {
+      throw new Error('Could not get core reservation id.');
+    }
+
+    const request = {
+      id: coreReservationId,
+      expiryDateTime: reserveNow.expiry_date.toISOString(),
+      idToken: {
+        idToken: reserveNow.token.uid,
+        type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+          reserveNow.token.type,
+        ),
+      },
+      evseId: evse.evseId,
+    } as ReserveNowRequest;
+    const storedCoreReservation =
+      await this.reservationRepo.createOrUpdateReservation(
+        request,
+        evse.stationId,
```

Comment:
It doesn't seem that `createOrUpdateReservation` takes in a stationId separately.

Reply:
I am afraid I did not get this point. Could you please share more for this point?

Reply:
@ChrisWeissmann i made the same assumption the first time -- this is referencing the core reservation repository, which **does** take in a station id, and not the ocpi reservation repository, which **does not** take in a station id, so it's all good, the arguments are correct!

@lydiazcheng this is what we get for having two similar repositories `reservationRepo` and `ocpiReservationRepo` haha, always get them confused 😄

Reply:
Gotcha. It's indeed confusing. I renamed the repo to `coreReservationRepo`.